File: Data-Structures/Trees/0103_zigzag_binary_tree.py
#trees
#fifo
#time complexity -> o(n), space complexity -> o(n) 
#use deque here to stop extra memory work -> if use list, popping lement results in empty slot and all elemnts have to shift
#deque is a doubly-linked list, so when pop elemnt, only that element is affected -> every single popleft is o(1) time

from typing import Optional 
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        # root, left to right, right to left, ... 
        if not root: 
            return []
        result = []
        queue = deque([root])
        left_to_right = True 

        while queue: 
            level_size = len(queue)
            current_level = [0] * level_size

            for i in range (level_size): 
                node = queue.popleft()
                #check if left or right, if right take the first right
                index = i if left_to_right else level_size - 1 - i 
                current_level[index] = node.val 
                if node.left: 
                    queue.append(node.left)
                if node.right: 
                    queue.append(node.right)
            
            result.append(current_level)
            left_to_right = not left_to_right

        return result

        

File: Data-Structures/Trees/0103_zigzag_binary_tree.c
// Practice for: 0103_zigzag_binary_tree.py
// Bridge logic from Python to C pointers here.

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h> 

struct TreeNode {
    int val; 
    struct TreeNode* left; 
    struct TreeNode* right;
};

int* zigzagLevelOrder(struct TreeNode* root, int* returnSize, int** returnColumnSizes){
    if (root == NULL){
        *returnSize = 0; 
        *returnColumnSizes = 0; 
        return NULL;
    }

    int** res = (int**)malloc(2000 * sizeof(int*)); 
    *returnColumnSizes = (int*)malloc(2000 * sizeof(int)); 
    *returnSize = 0; 

    struct TreeNode* queue[2000];
    int head = 0; 
    int tail = 0; 

    //place root here first
    queue[tail++] = root;
    int left_to_right = 1;

    while (head < tail){ 
        int levelSize = tail - head; 
        int* currentLevel = (int*)malloc(levelSize * sizeof(int)); 
        //col is top to bottom, row is left to right
        (*returnColumnSizes)[*returnSize] = levelSize;

        for (int i = 0; i < levelSize; i++){
            struct TreeNode* node = queue[head++];
            int index = left_to_right ? i : (levelSize - 1 - i);
            currentLevel[index] = node->val;

            if(node->left != NULL) queue[tail++] = node->left;
            if(node->right != NULL) queue[tail++] = node->right;
        }
        res[*returnSize] = currentLevel; 
        (*returnSize)++;
        left_to_right = !left_to_right;
    }
    return res;
}
